include/dali/core/mm/monotonic_resource.h

```diff
@@ -0,0 +1,203 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+#define DALI_CORE_MM_MONOTONIC_RESOURCE_H_
+
+#include <cassert>
+#include <algorithm>
+#include <utility>
+#include "dali/core/mm/memory_resource.h"
+#include "dali/core/mm/detail/util.h"
+#include "dali/core/mm/detail/align.h"
+#include "dali/core/small_vector.h"
+
+namespace dali {
+namespace mm {
+
+/**
+ * @brief Monotonic resource which returns memory from a preallocated buffer.
+ *
+ * Monotonic buffer resources don't require manual deallocation of the returned pointers.
+ * The memory is freed in bulk when the underlying buffer is freed.
+ */
+class monotonic_buffer_resource : public memory_resource {
+ public:
+  monotonic_buffer_resource() = default;
+  monotonic_buffer_resource(void *memory, size_t bytes)
+  : curr_(static_cast<char*>(memory)), limit_(static_cast<char*>(memory) + bytes) {}
+
+  size_t avail(size_t alignment) const {
+    return limit_ - detail::align_ptr(curr_, alignment);
+  }
+
+ protected:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_)
+      throw std::bad_alloc();
+    curr_ = ret + bytes;
+    return ret;
+  }
+
+  // don't deallocate at all
+  void do_deallocate(void *data, size_t bytes, size_t alignment) override {
+  }
+
+  char *curr_ = nullptr, *limit_ = nullptr;
+};
+
+/**
+ * @brief Monotonic resource which allocates memory from an upstream resource, storing
+ *        the metadata in the same memory blocks as the allocated buffers.
+ *
+ * Monotonic resources don't require manual deallocation of memory.
+ * The lifetime of a monotonic resource is limited and all memory will be deallocated in bulk
+ * when the resource is destroyed.
+ */
+class monotonic_host_resource : public memory_resource {
+ public:
+  explicit monotonic_host_resource(memory_resource *upstream,
+                                   size_t next_block_size = 1024)
+  : upstream_(upstream), next_block_size_(next_block_size) {}
+
+  ~monotonic_host_resource() {
+    free_all();
+  }
+
+  void free_all() {
+    while (curr_block_) {
+      assert(curr_block_->sentinel == sentinel_value && "Memory corruption detected");
+      auto *prev = curr_block_->prev;
+      auto *base = reinterpret_cast<char*>(curr_block_) - curr_block_->usable_size;
+      size_t alloc_size = curr_block_->usable_size + sizeof(block_info);
+      upstream_->deallocate(base, alloc_size, curr_block_->alignment);
+      curr_block_ = prev;
+    }
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_) {
+      while (next_block_size_ < bytes + sizeof(block_info))
+        next_block_size_ += next_block_size_;
```

Comment:
Do we want to double the next_block_size_  forever?

Reply:
Yes we do. That's why this is a _monotonic_ resource - it only ever grows.

Reply:
But do we want to double it every time or just byt a constant size?

Reply:
We want to double it.